File: backend/src/database/models/track.rs
use {
	bonsaidb::core::schema::Collection,
	chrono::NaiveDate,
	serde::{Deserialize, Serialize},
};

use crate::database::{models::InlinedArtist, views::track::TrackByReleaseId};

#[derive(Debug, Serialize, Deserialize, Collection)]
#[collection(name = "tracks", views = [TrackByReleaseId])]
pub struct Track {
	pub title: String,
	pub title_sort: Option<String>,
	pub track_number: Option<u32>,
	pub disc_number: Option<u32>,
	pub original_date: Option<NaiveDate>,

	/// Either [InlinedArtist::unknown] or populated with artists.
	pub artists: Vec<InlinedArtist>,
	pub artist_sort: Option<String>,

	/// Either [constants::UNKNOWN_RELEASE_ID] or a u64 of sorts.
	pub release_id: u64,
	pub composer_ids: Option<Vec<u64>>,
	pub producer_ids: Option<Vec<u64>>,
	pub cover_ids: Option<Vec<u64>>,

	pub genre_ids: Option<Vec<u64>>,
	pub tag_ids: Option<Vec<u64>>,

	pub mbz_id: Option<String>,
	pub path: String,
}

File: app/src/types/backend/track.ts
import type { InlinedArtist } from "@/types/backend/generic";
import type { Person } from "@/types/backend/person";

export type TrackCommand = "get_track_list_for_release";

export interface Track {
	title: string;
	title_sort: string | null;
	track_number: number | null;
	disc_number: number | null;
	original_date: string | null;

	artists: InlinedArtist[];
	artist_sort: string | null;

	release_id: number;
	composer_ids: number[] | null;
	producer_ids: number[] | null;
	cover_ids: number[] | null;

	genre_ids: number[] | null;
	tag_ids: number[] | null;

	mbz_id: string | null;
	path: string;
}

export interface DisplayTrackList {
	tracks: Track[];
	artists: Record<number, Person>;
}

export interface GetTrackListParameters {
	[key: string]: unknown;
	releaseId: number;
}
